---
title: "compare founder vars against a target bed"
author: "Ofer Isakov"
date: '2023-03-07'
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
setwd('/media/SSD/Bioinformatics/Projects/founder_variants_assessment_202204/')
library(ProjectTemplate)
load.project()
```

```{r definitions}
# target_file<-'./data/accessory_data/twist_hg38_exome_comp_spikein_v2.0.2_targets_sorted.re_annotated.pad50.bed'
# target_name<-'twist_exome_plus'
#target_file<-'/media/SSD/Bioinformatics/Databases/illumina/TruSight_Hereditary_Cancer_TargetedRegions_v2_hg19.pad50.sorted.bed'
target_name<-'trusight_hereditary_cancer'

output_dir<-glue('./output/variants_vs_target_bed/{target_name}')
dir.create(output_dir)
```

```{r functions}
parse_intersection_output<-function(intersect_output,founder_bed_prefix){
  num_original_variants<-nrow(intersect_output%>%distinct())
  not_covered_variants<-intersect_output%>%filter(is_covered==0)%>%
    mutate(source=founder_bed_prefix)%>%
    distinct()
  num_not_covered_variants<-nrow(not_covered_variants)
  num_intronic_variants_not_covered<-sum(grepl('+|-',not_covered_variants$hgvs_notation))
  message(glue('There are {num_not_covered_variants}/{num_original_variants} not covered variants. {num_intronic_variants_not_covered}/{num_not_covered_variants} are intronic'))
  return(not_covered_variants)
}

```

# MyScreen variants

```{r myscreen}
#founder_variants<-readr::read_delim('./output/founder_lists_with_genomic_coordinates/preconception_screening_myscreen_vep_202305.csv',delim='\t')
founder_variants<-readr::read_delim('./output/founder_lists_with_genomic_coordinates/preconception_screening_myscreen_vep.2024-03-08.csv',delim='\t')
founder_variants<-founder_variants%>%distinct()%>%select(chr,start,end,ref,alt,gene,variant,hgvs_notation)%>%mutate(chr=glue('chr{chr}'))
# generate bed file
founder_bed_prefix<-'preconception_screening_myscreen_vep_20240308'
founder_bed_file<-vep_to_bed(founder_variants,bed_prefix = founder_bed_prefix,add_chr = F)

intersect_output<-bedtools_intersect(founder_bed_file,b_file=target_file)
not_covered_variants<-parse_intersection_output(intersect_output,founder_bed_prefix)

founder_variants<-founder_variants%>%left_join(intersect_output)%>%mutate(source=founder_bed_prefix)
write.table(founder_variants,glue('{output_dir}/{target_name}_{founder_bed_prefix}_coverage.csv'),row.names = F,sep='\t',quote = F)

```

# MOH variants

```{r moh}
founder_variants<-readr::read_delim('./output/founder_lists_with_genomic_coordinates/preconception_screening_moh_vep.2024-03-06.csv',delim='\t')
founder_variants<-founder_variants%>%distinct()%>%select(chr,start,end,ref,alt,gene,variant,hgvs_notation)%>%mutate(chr=glue('chr{chr}'))
# generate bed file
founder_bed_prefix<-'moh_variants_20240306'
founder_bed_file<-vep_to_bed(founder_variants,bed_prefix = founder_bed_prefix)

intersect_output<-bedtools_intersect(founder_bed_file,b_file=target_file)
not_covered_variants<-parse_intersection_output(intersect_output,founder_bed_prefix)

founder_variants<-founder_variants%>%left_join(intersect_output%>%select(hgvs_notation,is_covered))%>%mutate(source=founder_bed_prefix)
write.table(founder_variants,glue('{output_dir}/{target_name}_{founder_bed_prefix}_not_covered.csv'),row.names = F,sep='\t',quote = F)

```

# Onco51 variants

```{r moh}
founder_variants<-readr::read_delim('./output/founder_lists_with_genomic_coordinates/onco51_founder_variants_vep.2024-03-11.csv',delim='\t')
founder_variants<-founder_variants%>%distinct()%>%select(chr,start,end,ref,alt,gene,variant,hgvs_notation)%>%mutate(chr=glue('chr{chr}'))
# generate bed file
founder_bed_prefix<-'onco51_panel_20240311'
founder_bed_file<-vep_to_bed(founder_variants,bed_prefix = founder_bed_prefix)

intersect_output<-bedtools_intersect(founder_bed_file,b_file=target_file)
not_covered_variants<-parse_intersection_output(intersect_output,founder_bed_prefix)

founder_variants<-founder_variants%>%left_join(intersect_output%>%select(hgvs_notation,is_covered))%>%mutate(source=founder_bed_prefix)
write.table(founder_variants,glue('{output_dir}/{target_name}_{founder_bed_prefix}_not_covered.csv'),row.names = F,sep='\t',quote = F)

```


# Merge variant files

```{r merge_files}
coverage_files<-list.files(output_dir,full.names = T)
all_coverage_variants<-NULL
for (nc_file in coverage_files){
  ncv<-readr::read_delim(nc_file,delim='\t')
  all_coverage_variants<-all_coverage_variants%>%bind_rows(ncv)
}

all_coverage_variants<-all_coverage_variants%>%filter(!is.na(start))

write.table(all_coverage_variants,
            file=glue('{output_dir}/{target_name}_founder_variants_not_covered.{Sys.Date()}.csv'),
            row.names = F,
            sep='\t',
            quote = F)
```