---
title: "vep variant parser"
author: "Ofer Isakov"
date: '2023-03-07'
output: html_document
---

# USE NOTES
- using vep_cds_to_genomic_coordinates - takes in an hgvs notation and returns the hg38 genomic coordinates
- if you want to convert from gene symbol with coordinates to genomic coordinates - you need to first find all the corresponding transcripts (using refseq2gene)
- there will be different transcripts with the same cds notation but with different genomic locations. the variant table should have a way to determine which is the correct one otherwise there is no way of knowing.


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
setwd('/media/SSD/Bioinformatics/Projects/founder_variants_assessment_202204/')
library(ProjectTemplate)
load.project()
```

```{r aa_conv}
amino_acids <- list(
  A = "Ala",
  R = "Arg",
  N = "Asn",
  D = "Asp",
  C = "Cys",
  E = "Glu",
  Q = "Gln",
  G = "Gly",
  H = "His",
  I = "Ile",
  L = "Leu",
  K = "Lys",
  M = "Met",
  F = "Phe",
  P = "Pro",
  S = "Ser",
  T = "Thr",
  W = "Trp",
  Y = "Tyr",
  V = "Val"
)

# Function to translate shorthand to long format
translate_amino_acids <- function(aa_change,protein_start) {
  # Split the input string by "/" to handle cases like "G/D"
  print(aa_change)
  if (is.na(aa_change)){
    return(data.frame(aa1=NA,aa2=NA,aa_change_long=NA))
  }
  split_sequence <- strsplit(aa_change, split = "/")[[1]]
  
  # Translate each amino acid in the sequence
  translated_sequence <- sapply(split_sequence, function(aa) {
    if (aa %in% names(amino_acids)) {
      return(amino_acids[[aa]])
    } else {
      return(aa) # Return the original character if no match is found
    }
  })
  aa_change_df<-data.frame(aa1=translated_sequence[1],aa2=translated_sequence[2],aa_change_long=glue('p.{translated_sequence[1]}{protein_start}{translated_sequence[2]}'))
  # Combine the translated sequence back into a string with "/"
  return(aa_change_df)
}

```

# MyScreen variants - revised

```{r definitions}

myscreen_variants<-readr::read_delim('./data/founder_variants_lists/davidov_et_al_clinical_genetics_s8.csv',delim='\t')
myscreen_for_vep<-myscreen_variants%>%
  mutate(gene=trimws(gene_name),variant=trimws(cds_seq))%>%
  select(-c(gene_name,cds_seq))%>%
  mutate(hgvs_notation=glue('{transcript_name}:{variant}'))%>%distinct()%>%
  filter(grepl('NM_',hgvs_notation))

# update transcript isoform 
# myscreen_for_vep<-myscreen_for_vep%>%
#   mutate(original_transcript_without_isoform=stringr::str_extract(transcript_name,'N[MRX]_[0-9]+'))%>%
#   left_join(refseq2gene%>%select(latest_isoform=X.name,original_transcript_without_isoform))%>%
#   mutate(transcript_name=latest_isoform)

# now fix problematic variants
# myscreen_for_vep<-
#   myscreen_for_vep%>%
#   mutate(hgvs_notation_fixed=
#            case_when(
#              hgvs_notation=='NM_015120.4:c.8008C>T'~'NM_001378454.1:c.8005C>T',
#              hgvs_notation=='NM_000186.3:c.3677_*4del'~'NM_000186.4:c.3677_*4del',
#              hgvs_notation=='NM_080680.2:c.3991C>T'~'NM_080680.3:c.3991C>T',
#              hgvs_notation=='NM_183380.3:c.14865delA'~'NM_183380.4:c.14865del',
#              hgvs_notation=='NM_000213.3:c.3280_3793+176del2279'~'NM_000213.5:c.3279_3793+180del',
#              hgvs_notation=='NM_002225.4:c.932C>T'~'NM_002225.5:c.932C>T',
#              hgvs_notation=='NM_018993.3:c.1731delC'~'NM_018993.4:c.1731del',
#              hgvs_notation=='NM_001164277.1:c.1179G>A'~'NM_001164277.2:c.1179G>A',
#              hgvs_notation=='NM_001164277.1:c.446G>A'~'NM_001164277.2:c.446G>A',
#              hgvs_notation=='NM_001164277.1:c.83G>A'~'NM_001164277.2:c.83G>A',
#              hgvs_notation=='NM_002420.5:c.-64+2213_899+118del36444'~'NM_002420.6:c.-64+2213_899+118del36444',
#              hgvs_notation=='NM_001127464.2:c.5943delA'~'NM_001367624.2:c.8627A>G',
#              !is.na(hgvs_notation)~hgvs_notation
#              )
#          )%>%
#   mutate(hgvs_notation=hgvs_notation_fixed)%>%select(-hgvs_notation_fixed)%>%
#   filter(!hgvs_notation%in%c('NM_153700.2:c.-78_5091+41del18645'))# filter out problematic variants

myscreen_for_vep_annotated<-NULL
for (i in 1:nrow(myscreen_for_vep)){
  message(glue('{i}/{nrow(myscreen_for_vep)}'))
  variant<-myscreen_for_vep%>%slice(i)
  variant_position<-vep_cds_to_genomic_coordinates(variant%>%pull(hgvs_notation),build='grch37')
  annotated_variant<-variant%>%left_join(variant_position)
  myscreen_for_vep_annotated<-myscreen_for_vep_annotated%>%bind_rows(annotated_variant)
}

validated_myscreen_for_vep_annotated<-myscreen_for_vep_annotated%>%filter(!grepl('HG|HS',chr))

write.table(validated_myscreen_for_vep_annotated%>%select(-c(consequence_terms,flags)),file=glue('./output/founder_lists_with_genomic_coordinates/preconception_screening_myscreen_vep.{Sys.Date()}.csv'),sep='\t',row.names=F,quote=F)

```

# Onco panel - revised
```{r onco_new}
onco51_variants<-readr::read_delim('./data/founder_variants_lists/onco_panel_51.csv',delim='\t')
onco51_variants<-onco51_variants%>%
  mutate(gene=trimws(gene),variant=trimws(variant),hgvs_notation=glue('{gene}:{variant}'))
onco51_for_vep<-onco51_variants%>%
  filter(hgvs_notation!='BMPR1A:c.-153+13146_*357832del')
  
onco51_for_vep_annotated<-NULL
for (i in 1:nrow(onco51_for_vep)){
  message(glue('{i}/{nrow(onco51_for_vep)}'))
  variant<-onco51_for_vep%>%slice(i)
  variant_position<-vep_cds_to_genomic_coordinates(variant%>%pull(hgvs_notation),build='grch37')
  annotated_variant<-variant%>%left_join(variant_position)
  onco51_for_vep_annotated<-onco51_for_vep_annotated%>%bind_rows(annotated_variant)
}

# validate onco51
# remove duplicate genomic regions
validated_onco51_for_vep_annotated<-onco51_for_vep_annotated%>%
  distinct(gene,variant,ethnicity,sector,chr,start,end,.keep_all = TRUE)%>%ungroup()
# grab the original cds position and compare to the cds_start
validated_onco51_for_vep_annotated<-validated_onco51_for_vep_annotated%>%
  mutate(original_cds_start=stringr::str_extract(variant,'\\d+'))%>%
  mutate(cds_start_match=ifelse(original_cds_start==cds_start,1,0),
         cds_start_match=ifelse(is.na(cds_start_match),0,cds_start_match))%>%
  group_by(gene,variant,ethnicity,sector)%>%
  slice_max(cds_start_match)%>%ungroup()
# grab the original protein position and compare to the protein_start
validated_onco51_for_vep_annotated<-validated_onco51_for_vep_annotated%>%
  mutate(original_protein_start=stringr::str_extract(protein,'\\d+'))%>%
  mutate(protein_start_match=ifelse(original_protein_start==protein_start,1,0),
         protein_start_match=ifelse(is.na(protein_start_match),0,protein_start_match))%>%
  group_by(gene,variant,ethnicity,sector)%>%
  slice_max(protein_start_match)%>%ungroup()
# grab the original first aa and compare to the annotation aa
validated_onco51_for_vep_annotated<-validated_onco51_for_vep_annotated%>%
  separate(amino_acids,sep='/',into=c('first_aa','second_aa'))%>%
  mutate(original_first_aa=stringr::str_extract(protein,'p.[^\\d]+')%>%str_replace('p.',''))%>%
  rowwise()%>%
  mutate(first_aa_translated=ifelse(first_aa %in% names(amino_acids),amino_acids[[first_aa]],NA))%>%
  mutate(first_aa_match=ifelse(original_first_aa==first_aa_translated,1,0),
         first_aa_match=ifelse(is.na(first_aa_match),0,first_aa_match))%>%
  group_by(gene,variant,ethnicity,sector)%>%
  slice_max(first_aa_match)%>%ungroup()
# if the variant is intronic (has +/- in the notation) then select the matching variant without cds
validated_onco51_for_vep_annotated<-validated_onco51_for_vep_annotated%>%
  mutate(original_is_intronic=ifelse(grepl('\\+|\\-',variant),1,0))%>%
  mutate(is_intronic=ifelse(is.na(cds_start),1,0))%>%
  mutate(match_is_intronic=ifelse(original_is_intronic==is_intronic,1,0))%>%
  group_by(gene,variant,ethnicity,sector)%>%
  slice_max(match_is_intronic)%>%ungroup()

# grab remaining duplicates
remaining_duplicates<-validated_onco51_for_vep_annotated%>%
  filter(duplicated(validated_onco51_for_vep_annotated%>%select(gene,variant,ethnicity,sector)))%>%select(hgvs_notation)%>%distinct()
# manually select the correct variants
validated_onco51_for_vep_annotated<-validated_onco51_for_vep_annotated%>%
  filter(!(hgvs_notation=='BRCA1:c.5266dup' & start!=41209081))%>% 
  filter(!(hgvs_notation=='BRCA1:c.5434C>G' & start!=41199693))

write.table(validated_onco51_for_vep_annotated%>%select(-consequence_terms),
            file=glue('./output/founder_lists_with_genomic_coordinates/onco51_founder_variants_vep.{Sys.Date()}.csv'),
            sep='\t',
            row.names=F,
            quote=F)

```

# MOH variants - revised

```{r definitions}

moh_variants<-readr::read_delim('./data/founder_variants_lists/moh_founder_unified.csv',delim='\t')
moh_variants<-moh_variants%>%
  mutate(gene=trimws(gene),variant=trimws(variant))
# convert problematic gene names
moh_for_vep<-moh_variants%>%
  mutate(
    gene = case_when(gene == 'C8orf37' ~ 'CFAP418',
                     gene == 'G6PC' ~ 'G6PC1',
                     !is.na(gene) ~ gene),
    hgvs_notation = glue('{gene}:{variant}'),
    hgvs_notation = case_when(
      hgvs_notation == 'TK2:c.542T>A' ~ 'TK2:c.635T>A',
      hgvs_notation == 'POR:c.1615G>A' ~ 'POR:c.1606G>A',
      hgvs_notation == 'CPS1:c.4101+2T>C' ~'CPS1:4134+2T>C',
      TRUE ~ hgvs_notation
    )
    
  )%>%
  filter(hgvs_notation!='SMN1:Exon7 del' & !variant%in%c('2p21del   [NC_000002.11:41727140_47727139del]',
                                                         'Del exons 3 and 4'))

moh_for_vep_annotated<-NULL
for (i in 1:nrow(moh_for_vep)){
  message(glue('{i}/{nrow(moh_for_vep)}'))
  variant<-moh_for_vep%>%slice(i)
  variant_position<-vep_cds_to_genomic_coordinates(variant%>%pull(hgvs_notation),build='grch37')
  annotated_variant<-variant%>%left_join(variant_position)
  moh_for_vep_annotated<-moh_for_vep_annotated%>%bind_rows(annotated_variant)
}

compare_with_original<-function(original_table,modified_table){
  missing_variants<-NULL
  for (i in 1:nrow(original_table)){
    original_variant<-original_table%>%slice(i)
    if (!original_variant%>%pull(hgvs_notation) %in% modified_table$hgvs_notation){
      missing_variants<-missing_variants%>%bind_rows(original_variant)
    }else{
      message(glue('found {original_variant%>%pull(hgvs_notation)}'))
    }
  }
  return(missing_variants)
}

compare_with_original(moh_for_vep,moh_for_vep_annotated)

# validate
# remove duplicate genomic regions
validated_moh_for_vep_annotated<-moh_for_vep_annotated%>%
  distinct(gene,hgvs_notation,ethnicity,chr,start,end,.keep_all = TRUE)
# compare annotation hgvs with variant annotation
validated_moh_for_vep_annotated<-validated_moh_for_vep_annotated%>%
  mutate(original_hgvs_cds=stringr::str_extract(variant,'c\\..+')%>%stringr::str_replace('del.+','del')%>%stringr::str_replace('dup.+','dup'),
         annotation_hgvs_cds=stringr::str_extract(hgvsc,'c\\..+'))%>%
  mutate(cds_start_match=ifelse(original_hgvs_cds==annotation_hgvs_cds,1,0),
         cds_start_match=ifelse(is.na(cds_start_match),0,cds_start_match))%>%
  group_by(gene,variant,ethnicity,locality)%>%
  slice_max(cds_start_match)%>%ungroup()

# grab the original cds position and compare to the cds_start
validated_moh_for_vep_annotated<-validated_moh_for_vep_annotated%>%
  mutate(original_cds_start=stringr::str_extract(variant,'\\d+'))%>%
  mutate(cds_start_match=ifelse(original_cds_start==cds_start,1,0),
         cds_start_match=ifelse(is.na(cds_start_match),0,cds_start_match))%>%
  group_by(gene,variant,ethnicity,locality)%>%
  slice_max(cds_start_match)%>%ungroup()

# if the variant is intronic (has +/- in the notation) then select the matching variant without cds
validated_moh_for_vep_annotated<-validated_moh_for_vep_annotated%>%
  mutate(original_is_intronic=ifelse(grepl('\\+|\\-',variant),1,0))%>%
  mutate(is_intronic=ifelse(is.na(cds_start),1,0))%>%
  mutate(match_is_intronic=ifelse(original_is_intronic==is_intronic,1,0))%>%
  group_by(gene,variant,ethnicity,locality)%>%
  slice_max(match_is_intronic)%>%ungroup()

# manually select the correct variants
validated_moh_for_vep_annotated<-validated_moh_for_vep_annotated%>%
  filter(!(hgvs_notation=='PGM1:c.112A>T' & start!=64059271))%>%
  filter(!(hgvs_notation=='HEXA:c.835T>C' & start!=72641571))%>%
  filter(!(hgvs_notation=='CPS1:c.3392C>T' & start!=211513252))%>%
  filter(!(hgvs_notation=='PCDH15:c.733C>T' & start!=56077174))%>%
  filter(!(hgvs_notation=='POR:c.1606G>A' & start!=75615113))

remaining_duplicates<-validated_moh_for_vep_annotated%>%
  filter(duplicated(validated_moh_for_vep_annotated%>%select(gene,hgvs_notation,ethnicity,locality)))%>%distinct()



write.table(validated_moh_for_vep_annotated%>%select(-consequence_terms),
            file=glue('./output/founder_lists_with_genomic_coordinates/preconception_screening_moh_vep.{Sys.Date()}.csv'),
            sep='\t',
            row.names=F,
            quote=F)

```


# Zlotogora

```{r zlotogora}

zlotogora_variants<-xlsx::read.xlsx('./data/founder_variants_lists/zlotogora_ingd_202203.xls',sheetIndex = 1)
# filter only founders
zlotogora_founder_variants<-zlotogora_variants%>%filter(grepl('founde',X.No.of.chromosomes.families,ignore.case=T))%>%
  mutate(variant=Mutation.c)%>%mutate(gene=trimws(Gene),variant=trimws(variant))
zlotogora_for_vep<-zlotogora_founder_variants%>%
  left_join(refseq2gene%>%filter(!grepl('_',chrom))%>%distinct(),by=c('gene'='name2'))%>%
  rename(transcript_name=X.name)%>%
  filter(!is.na(transcript_name))%>%filter(!is.na(variant))%>%filter(grepl('^c.',variant))%>%
  mutate(variant=stringr::str_replace(variant,'\\;.+',''))%>%mutate(variant=stringr::str_replace_all(variant,'\\s+',''))%>%
  mutate(hgvs_notation=glue('{transcript_name}:{variant}'))%>%distinct()

zlotogora_for_vep_annotated<-NULL
for (i in 1:nrow(zlotogora_for_vep)){
  variant<-zlotogora_for_vep%>%slice(i)
  variant_position<-vep_cds_to_genomic_coordinates(variant%>%pull(hgvs_notation))
  annotated_variant<-variant%>%left_join(variant_position)
  zlotogora_for_vep_annotated<-zlotogora_for_vep_annotated%>%bind_rows(annotated_variant)
}


final_zlotogora_variants_with_annotation<-zlotogora_for_vep_annotated%>%filter(!is.na(seq_region_name))
# add grch37 coordinates in order to validate the variants
final_zlotogora_variants_with_annotation_and_grch37<-NULL
for (i in 1:nrow(final_zlotogora_variants_with_annotation)){
  var<-final_zlotogora_variants_with_annotation%>%slice(i)%>%
      mutate(var_coords=glue('{seq_region_name}:{start}..{end}'))
  converted_coords<-vep_convert_coordinates('GRCh38','GRCh37',coordinates = var%>%pull(var_coords))
  var<-var%>%bind_cols(converted_coords)
  final_zlotogora_variants_with_annotation_and_grch37<-final_zlotogora_variants_with_annotation_and_grch37%>%bind_rows(var)
  message(glue('converted: {var%>%pull(var_coords)} : {var%>%pull(chr_GRCh37)}:{var%>%pull(start_GRCh37)}..{var%>%pull(end_GRCh37)}'))
}

# finally, where available, validate the coordinates in the location column (original zlotogora table) match the ones in the GRCh37 
# if no location is available, keep the row
final_zlotogora_variants_with_annotation_and_grch37<-final_zlotogora_variants_with_annotation_and_grch37%>%mutate(
  match=location==glue('chr{chr_GRCh37}:{start_GRCh37}')
)

final_zlotogora_variants_with_annotation_and_grch37_clean<-final_zlotogora_variants_with_annotation_and_grch37%>%
  filter(is.na(match)|match)%>%
  select(
  population,Ethnic.group,Gene,hgvs_notation,
  chr_GRCh38=seq_region_name,start_GRCh38=start,end_GRCh38=end,
  chr_GRCh37,start_GRCh37,end_GRCh37,ref,alt
)%>%
  distinct(population,Ethnic.group,Gene,chr_GRCh38,start_GRCh38,end_GRCh38,.keep_all = T)

write.table(final_zlotogora_variants_with_annotation_and_grch37_clean,
            file=glue('./output/founder_lists_with_genomic_coordinates/zlotogora_founder_variants_vep.{Sys.Date()}.csv'),
            sep='\t',
            row.names=F,
            quote=F)

```

```{r add_grch37_coordinates}
assemb_from<-'GRCh38'
assemb_to<-'GRCh37'

input_files<-c(#'./output/founder_lists_with_genomic_coordinates/preconception_screening_moh_vep.2023-10-04.csv',
               #'./output/founder_lists_with_genomic_coordinates/preconception_screening_myscreen_vep.2023-10-04.csv',
               './output/founder_lists_with_genomic_coordinates/onco51_panel_vep.2023-10-05.csv')


for (input_file in input_files){
  message(glue('Annotating {input_file}'))
  input_vars<-readr::read_delim(input_file,delim='\t',na = 'NA')
  output_vars<-NULL
  for (i in 1:nrow(input_vars)){
    var<-input_vars%>%slice(i)%>%
      mutate(var_coords=glue('{chr}:{start}..{end}'))
    if (!is.na(var%>%pull(start))){
      message(glue('converting {var%>%pull(var_coords)}'))
      converted_coords<-vep_convert_coordinates(assemb_from,assemb_to,coordinates = var%>%pull(var_coords))
      var<-var%>%bind_cols(converted_coords)
    }
    output_vars<-output_vars%>%bind_rows(var)
  }
  output_file_name<-glue("{stringr::str_replace(basename(input_file),'.csv','')}.with{assemb_to}.csv")
  write.table(output_vars,file=glue('./output/founder_lists_with_genomic_coordinates/{output_file_name}'),row.names = F,sep='\t')
}
```

# clinvar
```{r oncopanel}
library(vcfR)
library(dplyr)
clinvar_file_vcf<-vcfR::read.vcfR('/media/SSD/Bioinformatics/Databases/clinvar/clinvar_hg38_20231028.plp.vcf')
clinvar_file_vcf_tidy<-vcfR2tidy(clinvar_file_vcf,info_only = T)
clinvar_fixed<-clinvar_file_vcf_tidy$fix%>%separate_rows(GENEINFO,sep = '\\|')%>%
  separate(GENEINFO,into=c('gene_symbol','gene_id'),sep = ':',remove = F)

#onco51_for_vep<-onco51_for_vep%>%filter(gene%in%(bad_vars%>%filter(!is.na(transcript_name))%>%pull(gene)))
query_size<-125

var_chunks<-split(clinvar_fixed$CLNHGVS, ceiling(seq_along(clinvar_fixed$CLNHGVS)/query_size))
chunk_num<-1
clinvar_all_annotated_vars<-NULL

for (hgvs_notations in var_chunks[chunk_num:length(var_chunks)]){
  message(glue('RUNNING CHUNK:{chunk_num}'))
  annotated_vars<-vep_basic_annotation(hgvs_notations)
  #annotated_vars<-validate_vep_variants(annotated_vars,keep=T)
  clinvar_all_annotated_vars<-clinvar_all_annotated_vars%>%bind_rows(annotated_vars)
  chunk_num<-chunk_num+1
}

save(clinvar_all_annotated_vars,file = './output/clinvar_annotated.RData')

all_var_freqs<-NULL
for (var_colocated in clinvar_all_annotated_vars$colocated_variants){
  #print(var_colocated)
  if ('frequencies' %in% colnames(var_colocated)){
    var_freqs<-var_colocated%>%pull(frequencies)
    var_freqs<-var_freqs[[1]]%>%na.omit()
    #print(var_freqs)
    all_var_freqs<-all_var_freqs%>%bind_rows(var_freqs)
  }else{
    all_var_freqs<-all_var_freqs%>%bind_rows(data.frame(af=NA))
  }
}
clinvar_with_freq<-clinvar_fixed%>%left_join(clinvar_all_annotated_vars%>%select(CLNHGVS=input)%>%bind_cols(all_var_freqs))
write.table(clinvar_with_freq,file='/media/SSD/Bioinformatics/Projects/alpha_missense_app_202309/data/clinvar.hg38.plp.with_freq.csv',sep='\t',row.names = F)
```

# Compare variant files
```{r compare_files}
input_files<-c('./output/founder_lists_with_genomic_coordinates/preconception_screening_moh_vep.2023-10-04.withGRCh37.csv',
               './output/founder_lists_with_genomic_coordinates/preconception_screening_myscreen_vep.2023-10-04.csv',
               './output/founder_lists_with_genomic_coordinates/onco51_panel_vep.2023-10-05.csv',
               './output/founder_lists_with_genomic_coordinates/zlotogora_founder_variants_vep.2024-02-29.csv')


for (ifile in input_files){
  fv_data<-readr::read_delim(ifile,delim='\t')
  print(colnames(fv_data))
}

```
