---
title: "vep variant parser"
author: "Ofer Isakov"
date: '2023-03-07'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
setwd('/media/SSD/Bioinformatics/Projects/founder_variants_assessment_202204/')
library(ProjectTemplate)
load.project()
```

```{r aa_conv}
amino_acids <- list(
  A = "Ala",
  R = "Arg",
  N = "Asn",
  D = "Asp",
  C = "Cys",
  E = "Glu",
  Q = "Gln",
  G = "Gly",
  H = "His",
  I = "Ile",
  L = "Leu",
  K = "Lys",
  M = "Met",
  F = "Phe",
  P = "Pro",
  S = "Ser",
  T = "Thr",
  W = "Trp",
  Y = "Tyr",
  V = "Val"
)

# Function to translate shorthand to long format
translate_amino_acids <- function(aa_change,protein_start) {
  # Split the input string by "/" to handle cases like "G/D"
  print(aa_change)
  if (is.na(aa_change)){
    return(data.frame(aa1=NA,aa2=NA,aa_change_long=NA))
  }
  split_sequence <- strsplit(aa_change, split = "/")[[1]]
  
  # Translate each amino acid in the sequence
  translated_sequence <- sapply(split_sequence, function(aa) {
    if (aa %in% names(amino_acids)) {
      return(amino_acids[[aa]])
    } else {
      return(aa) # Return the original character if no match is found
    }
  })
  aa_change_df<-data.frame(aa1=translated_sequence[1],aa2=translated_sequence[2],aa_change_long=glue('p.{translated_sequence[1]}{protein_start}{translated_sequence[2]}'))
  # Combine the translated sequence back into a string with "/"
  return(aa_change_df)
}

```

# MyScreen variants - revised

```{r definitions}

myscreen_variants<-readr::read_delim('./data/founder_variants_lists/davidov_et_al_clinical_genetics_s8.csv',delim='\t')
myscreen_for_vep<-myscreen_variants%>%
  mutate(gene=trimws(gene_name),variant=trimws(cds_seq))%>%
  select(-c(gene_name,cds_seq))%>%
  mutate(hgvs_notation=glue('{transcript_name}:{variant}'))%>%distinct()%>%
  filter(grepl('NM_',hgvs_notation))

myscreen_for_vep_annotated<-NULL
for (i in 1:nrow(myscreen_for_vep)){
#for (i in 1:100){
  message(glue('{i}/{nrow(myscreen_for_vep)}'))
  variant<-myscreen_for_vep%>%slice(i)
  variant_position<-vep_cds_to_genomic_coordinates(variant%>%pull(hgvs_notation),build='grch37',gene_symbol_or_refseq = 'refseq')
  if ((nrow(variant_position)>0) &&(!('error' %in% colnames(variant_position))) ){
    annotated_variant<-variant%>%left_join(variant_position)
  }else{ # if the transcript was not found, use the gene name instead
    variant_position<-vep_cds_to_genomic_coordinates(glue('{variant%>%pull(gene)}:{variant%>%pull(variant)}'),build='grch37',gene_symbol_or_refseq = 'gene_symbol')
    if ((nrow(variant_position)>0) &&(!('error' %in% colnames(variant_position))) ){
      annotated_variant<-variant%>%select(-hgvs_notation)%>%bind_cols(variant_position%>%mutate(hgvs_notation=hgvsc))
    }else{
      message('could not parse the variant both by transcript and by gene')
      annotated_variant<-variant
    }
  }
  myscreen_for_vep_annotated<-myscreen_for_vep_annotated%>%bind_rows(annotated_variant)
}

validated_myscreen_for_vep_annotated<-myscreen_for_vep_annotated%>%filter(!grepl('HG|HS',chr))%>%distinct(hgvsc,.keep_all = T)
# validate CDS
validated_myscreen_for_vep_annotated<-validated_myscreen_for_vep_annotated%>%
  mutate(original_hgvs_cds=stringr::str_extract(variant,'c\\..+')%>%stringr::str_replace('del.+','del')%>%
           stringr::str_replace('dup.+','dup'),
         annotation_hgvs_cds=stringr::str_extract(hgvsc,'c\\..+'))%>%
  mutate(cds_start_match=ifelse(original_hgvs_cds==annotation_hgvs_cds,1,0),
         cds_start_match=ifelse(is.na(cds_start_match),0,cds_start_match))%>%
  group_by(transcript_name,variant)%>%
  slice_max(cds_start_match)%>%ungroup()
# validate Protein seq
validated_myscreen_for_vep_annotated<-validated_myscreen_for_vep_annotated%>%
  mutate(original_p=stringr::str_replace_all(protein_seq,'p\\.\\(|\\)','')%>%stringr::str_replace('\\d+',''),
         annotation_p=stringr::str_replace_all(hgvsp,'.+p\\.','')%>%stringr::str_replace('Ter','*')%>%stringr::str_replace('\\d+',''))%>%
  mutate(p_match=ifelse(original_p==annotation_p,1,0),
         p_match=ifelse(is.na(p_match),0,p_match))%>%
  group_by(transcript_name,variant)%>%
  slice_max(p_match)%>%ungroup()
# pick cannonical
validated_myscreen_for_vep_annotated<-validated_myscreen_for_vep_annotated%>%
  group_by(transcript_name,variant)%>%
  slice_max(canonical)%>%ungroup()

# remove ambiguous variants
ambiguous_variants<-validated_myscreen_for_vep_annotated%>%select(gene,variant)%>%filter(duplicated(validated_myscreen_for_vep_annotated%>%select(gene,variant)))%>%distinct()
validated_myscreen_for_vep_annotated<-validated_myscreen_for_vep_annotated%>%
  filter(!(gene%in%ambiguous_variants$gene & variant%in%ambiguous_variants$variant))


write.table(validated_myscreen_for_vep_annotated%>%select(-c(consequence_terms,flags)),file=glue('./output/founder_lists_with_genomic_coordinates/preconception_screening_myscreen_vep.{Sys.Date()}.csv'),sep='\t',row.names=F,quote=F)

```

# Onco panel - revised
```{r onco_new}
onco51_variants<-readr::read_delim('./data/founder_variants_lists/onco_panel_51.csv',delim='\t')
onco51_variants<-onco51_variants%>%
  mutate(gene=trimws(gene),variant=trimws(variant),hgvs_notation=glue('{gene}:{variant}'))
onco51_for_vep<-onco51_variants%>%
  filter(hgvs_notation!='BMPR1A:c.-153+13146_*357832del')
  
onco51_for_vep_annotated<-NULL
for (i in 1:nrow(onco51_for_vep)){
  message(glue('{i}/{nrow(onco51_for_vep)}'))
  variant<-onco51_for_vep%>%slice(i)
  variant_position<-vep_cds_to_genomic_coordinates(variant%>%pull(hgvs_notation),build='grch37',gene_symbol_or_refseq = 'gene_symbol')
  annotated_variant<-variant%>%left_join(variant_position)
  onco51_for_vep_annotated<-onco51_for_vep_annotated%>%bind_rows(annotated_variant)
}

# validate onco51
# remove duplicate genomic regions
validated_onco51_for_vep_annotated<-onco51_for_vep_annotated%>%
  distinct(gene,variant,ethnicity,sector,chr,start,end,.keep_all = TRUE)%>%ungroup()
# validate cds notation
validated_onco51_for_vep_annotated<-validated_onco51_for_vep_annotated%>%
  mutate(original_hgvs_cds=stringr::str_replace(variant,'del.+','del')%>%stringr::str_replace('dup.+','dup'),
         annotation_hgvs_cds=stringr::str_extract(hgvsc,'c\\..+'))%>%
  mutate(cds_start_match=ifelse(original_hgvs_cds==annotation_hgvs_cds,1,0),
         cds_start_match=ifelse(is.na(cds_start_match),0,cds_start_match))%>%
  group_by(gene,variant,ethnicity,sector)%>%
  slice_max(cds_start_match)%>%ungroup()
# compare start protein
validated_onco51_for_vep_annotated<-validated_onco51_for_vep_annotated%>%
  mutate(original_protein=stringr::str_extract(protein,'p\\..+')%>%stringr::str_replace('\\d+.+',''),
         annotation_protein=stringr::str_extract(hgvsp,'p\\..+')%>%stringr::str_replace('\\d+.+',''))%>%
  mutate(protein_match=ifelse(original_protein==annotation_protein,1,0),
         protein_match=ifelse(is.na(protein_match),0,protein_match))%>%
  group_by(gene,variant,ethnicity,sector)%>%
  slice_max(protein_match)%>%ungroup()
# compare protein notation
validated_onco51_for_vep_annotated<-validated_onco51_for_vep_annotated%>%
  mutate(annotation_protein=stringr::str_extract(hgvsp,'p\\..+'))%>%
  mutate(protein_match=ifelse(protein==annotation_protein,1,0),
         protein_match=ifelse(is.na(protein_match),0,protein_match))%>%
  group_by(gene,variant,ethnicity,sector)%>%
  slice_max(protein_match)%>%ungroup()
# if the variant is intronic (has +/- in the notation) then select the matching variant without cds
validated_onco51_for_vep_annotated<-validated_onco51_for_vep_annotated%>%
  mutate(original_is_intronic=ifelse(grepl('\\+|\\-',variant),1,0))%>%
  mutate(is_intronic=ifelse(is.na(cds_start),1,0))%>%
  mutate(match_is_intronic=ifelse(original_is_intronic==is_intronic,1,0))%>%
  group_by(gene,variant,ethnicity,sector)%>%
  slice_max(match_is_intronic)%>%ungroup()
# fix deletions
validated_onco51_for_vep_annotated<-validated_onco51_for_vep_annotated%>%
  mutate(hgvs_del_ref=stringr::str_extract(hgvs_notation,'del.+'),
         annotation_del_ref=glue('del{ref}'))%>%
  mutate(del_match=ifelse(hgvs_del_ref==annotation_del_ref,1,0),
         del_match=ifelse(is.na(del_match),0,del_match))%>%
  group_by(gene,variant,ethnicity,sector)%>%
  slice_max(del_match)%>%ungroup()

# grab remaining duplicates
remaining_duplicates<-validated_onco51_for_vep_annotated%>%
  filter(duplicated(validated_onco51_for_vep_annotated%>%select(gene,variant,ethnicity,sector)))%>%select(hgvs_notation)%>%distinct()
# manually select the correct variants
validated_onco51_for_vep_annotated<-validated_onco51_for_vep_annotated%>%
  filter(!(hgvs_notation=='BRCA1:c.5266dup' & start!=41209081))%>% 
  filter(!(hgvs_notation=='BRCA1:c.5434C>G' & start!=41199693))%>%
  filter(!(hgvs_notation=='BRCA1:c.5123C>A' & start!=41215920))%>%
  filter(!(hgvs_notation=='CHEK2:c.499G>A' & start!=29121058))%>%
  filter(!(hgvs_notation=='MSH6:c.3603_3606delCATG' & start!=48032803))

write.table(validated_onco51_for_vep_annotated%>%select(-consequence_terms),
            file=glue('./output/founder_lists_with_genomic_coordinates/onco51_founder_variants_vep.{Sys.Date()}.csv'),
            sep='\t',
            row.names=F,
            quote=F)

```

# MOH variants - revised

```{r definitions}

moh_variants<-readr::read_delim('./data/founder_variants_lists/moh_founder_unified.csv',delim='\t')
moh_variants<-moh_variants%>%
  mutate(gene=trimws(gene),variant=trimws(variant))
# convert problematic gene names
moh_for_vep<-moh_variants%>%
  mutate(
    gene = case_when(gene == 'C8orf37' ~ 'CFAP418',
                     gene == 'G6PC' ~ 'G6PC1',
                     !is.na(gene) ~ gene),
    hgvs_notation = glue('{gene}:{variant}'),
    hgvs_notation = case_when(
      hgvs_notation == 'TK2:c.542T>A' ~ 'TK2:c.635T>A',
      hgvs_notation == 'POR:c.1615G>A' ~ 'POR:c.1606G>A',
      hgvs_notation == 'CPS1:c.4101+2T>C' ~'CPS1:4134+2T>C',
      TRUE ~ hgvs_notation
    )
    
  )%>%
  filter(hgvs_notation!='SMN1:Exon7 del' & !variant%in%c('2p21del   [NC_000002.11:41727140_47727139del]',
                                                         'Del exons 3 and 4'))

moh_for_vep_annotated<-NULL
for (i in 1:nrow(moh_for_vep)){
  message(glue('{i}/{nrow(moh_for_vep)}'))
  variant<-moh_for_vep%>%slice(i)
  variant_position<-vep_cds_to_genomic_coordinates(variant%>%pull(hgvs_notation),build='grch37',gene_symbol_or_refseq = 'gene_symbol')
  annotated_variant<-variant%>%left_join(variant_position)
  moh_for_vep_annotated<-moh_for_vep_annotated%>%bind_rows(annotated_variant)
}

# validate
# remove duplicate genomic regions
validated_moh_for_vep_annotated<-moh_for_vep_annotated%>%
  distinct(gene,hgvs_notation,ethnicity,chr,start,end,.keep_all = TRUE)
# compare annotation hgvs with variant annotation
validated_moh_for_vep_annotated<-validated_moh_for_vep_annotated%>%
  mutate(original_hgvs_cds=stringr::str_extract(variant,'c\\..+')%>%stringr::str_replace('del.+','del')%>%stringr::str_replace('dup.+','dup'),
         annotation_hgvs_cds=stringr::str_extract(hgvsc,'c\\..+'))%>%
  mutate(cds_start_match=ifelse(original_hgvs_cds==annotation_hgvs_cds,1,0),
         cds_start_match=ifelse(is.na(cds_start_match),0,cds_start_match))%>%
  group_by(gene,variant,ethnicity,locality)%>%
  slice_max(cds_start_match)%>%ungroup()

# grab the original cds position and compare to the cds_start
validated_moh_for_vep_annotated<-validated_moh_for_vep_annotated%>%
  mutate(original_cds_start=stringr::str_extract(variant,'\\d+'))%>%
  mutate(cds_start_match=ifelse(original_cds_start==cds_start,1,0),
         cds_start_match=ifelse(is.na(cds_start_match),0,cds_start_match))%>%
  group_by(gene,variant,ethnicity,locality)%>%
  slice_max(cds_start_match)%>%ungroup()

# if the variant is intronic (has +/- in the notation) then select the matching variant without cds
validated_moh_for_vep_annotated<-validated_moh_for_vep_annotated%>%
  mutate(original_is_intronic=ifelse(grepl('\\+|\\-',variant),1,0))%>%
  mutate(is_intronic=ifelse(is.na(cds_start),1,0))%>%
  mutate(match_is_intronic=ifelse(original_is_intronic==is_intronic,1,0))%>%
  group_by(gene,variant,ethnicity,locality)%>%
  slice_max(match_is_intronic)%>%ungroup()

# fix deletions
validated_moh_for_vep_annotated<-validated_moh_for_vep_annotated%>%
  mutate(hgvs_del_ref=stringr::str_extract(hgvs_notation,'del.+'),
         annotation_del_ref=glue('del{ref}'))%>%
  mutate(del_match=ifelse(hgvs_del_ref==annotation_del_ref,1,0),
         del_match=ifelse(is.na(del_match),0,del_match))%>%
  group_by(gene,variant,ethnicity,locality)%>%
  slice_max(del_match)%>%ungroup()

# grab remaining duplicates
remaining_duplicates<-validated_moh_for_vep_annotated%>%
  filter(duplicated(validated_moh_for_vep_annotated%>%select(gene,variant,ethnicity,locality)))%>%select(hgvs_notation)%>%distinct()

# remove ambiguous
validated_moh_for_vep_annotated<-validated_moh_for_vep_annotated%>%filter(!hgvs_notation%in%remaining_duplicates$hgvs_notation)

write.table(validated_moh_for_vep_annotated%>%select(-consequence_terms),
            file=glue('./output/founder_lists_with_genomic_coordinates/preconception_screening_moh_vep.{Sys.Date()}.csv'),
            sep='\t',
            row.names=F,
            quote=F)

```


# Zlotogora

```{r zlotogora}

#zlotogora_variants<-xlsx::read.xlsx('./data/founder_variants_lists/zlotogora_ingd_202203.xls',sheetIndex = 1)
zlotogora_variants<-readr::read_delim('./data/founder_variants_lists/zlotogora_ignd_20240329.csv',delim='\t',comment='')
colnames(zlotogora_variants)<-make.names(colnames(zlotogora_variants))
# filter only founders
zlotogora_founder_variants<-zlotogora_variants%>%
  #filter(grepl('founde',X.No.of.chromosomes.families,ignore.case=T))%>% # add this if you want to take only the variants that are marked as founder
  mutate(Mutation.c=stringr::str_replace_all(Mutation.c,'−','-'),
         Mutation.c=stringr::str_replace_all(Mutation.c,'→','>'))%>%
  mutate(variant=Mutation.c)%>%mutate(gene=trimws(Gene),variant=trimws(variant))%>%
  mutate(gene=stringr::str_replace_all(gene,'\\s+',''))%>%
  mutate(variant=stringr::str_replace_all(variant,'‐','-'))
zlotogora_for_vep<-zlotogora_founder_variants%>%
  filter(!is.na(variant))%>%filter(grepl('^c.',variant))%>%
  mutate(variant=stringr::str_replace(variant,'\\;.+',''))%>%mutate(variant=stringr::str_replace_all(variant,'\\s+',''))%>%
  mutate(hgvs_notation=glue('{gene}:{variant}'))%>%distinct()

zlotogora_for_vep_annotated<-NULL
for (i in 1:nrow(zlotogora_for_vep)){
#for (i in 3171:nrow(zlotogora_for_vep)){
  message(glue('{i}/{nrow(zlotogora_for_vep)}'))
  variant<-zlotogora_for_vep%>%slice(i)
  variant_position<-vep_cds_to_genomic_coordinates(variant%>%pull(hgvs_notation),build='grch37',gene_symbol_or_refseq = 'gene_symbol')
  annotated_variant<-variant%>%left_join(variant_position)
  zlotogora_for_vep_annotated<-zlotogora_for_vep_annotated%>%bind_rows(annotated_variant)
}

# validate variants
validated_zlotogora_variants_with_annotation<-zlotogora_for_vep_annotated%>%filter(!is.na(chr))
# remove duplicate genomic regions
validated_zlotogora_variants_with_annotation<-validated_zlotogora_variants_with_annotation%>%
  distinct(gene,variant,Ethnic.group,locality,chr,start,end,.keep_all = TRUE)
# if there is a location, validate it
validated_zlotogora_variants_with_annotation<-validated_zlotogora_variants_with_annotation%>%
  mutate(annotation_location=glue('chr{chr}:{start}'))%>%
  mutate(location_match=ifelse(location==annotation_location,1,0),
         location_match=ifelse(is.na(location_match),0,location_match))%>%
  group_by(gene,variant,Ethnic.group,locality)%>%
  slice_max(location_match)%>%ungroup()
# compare annotation hgvs with variant annotation
validated_zlotogora_variants_with_annotation<-validated_zlotogora_variants_with_annotation%>%
  mutate(original_hgvs_cds=stringr::str_extract(variant,'c\\..+')%>%stringr::str_replace('del.+','del')%>%stringr::str_replace('dup.+','dup'),
         annotation_hgvs_cds=stringr::str_extract(hgvsc,'c\\..+'))%>%
  mutate(cds_start_match=ifelse(original_hgvs_cds==annotation_hgvs_cds,1,0),
         cds_start_match=ifelse(is.na(cds_start_match),0,cds_start_match))%>%
  group_by(gene,variant,Ethnic.group,locality)%>%
  slice_max(cds_start_match)%>%ungroup()
# compare annotation protein with variant annotation
validated_zlotogora_variants_with_annotation<-validated_zlotogora_variants_with_annotation%>%
  mutate(annotation_protein=stringr::str_extract(hgvsp,'p\\..+'))%>%
  mutate(protein_match=ifelse(Mutation..p.==annotation_protein,1,0),
         protein_match=ifelse(is.na(protein_match),0,protein_match))%>%
  group_by(gene,variant,Ethnic.group,locality)%>%
  slice_max(protein_match)%>%ungroup()

# fix deletions
validated_zlotogora_variants_with_annotation<-validated_zlotogora_variants_with_annotation%>%
  mutate(hgvs_del_ref=stringr::str_extract(hgvs_notation,'del.+'),
         annotation_del_ref=glue('del{ref}'))%>%
  mutate(del_match=ifelse(hgvs_del_ref==annotation_del_ref,1,0),
         del_match=ifelse(is.na(del_match),0,del_match))%>%
  group_by(gene,variant,Ethnic.group,locality)%>%
  slice_max(del_match)%>%ungroup()

# fix duplications
validated_zlotogora_variants_with_annotation<-validated_zlotogora_variants_with_annotation%>%
  mutate(hgvs_del_ref=stringr::str_extract(hgvs_notation,'dup.+'),
         annotation_del_ref=glue('dup{alt}'))%>%
  mutate(dup_match=ifelse(hgvs_del_ref==annotation_del_ref,1,0),
         dup_match=ifelse(is.na(dup_match),0,dup_match))%>%
  group_by(gene,variant,Ethnic.group,locality)%>%
  slice_max(dup_match)%>%ungroup()

# remove same pos variants
validated_zlotogora_variants_with_annotation<-validated_zlotogora_variants_with_annotation%>%
  distinct(gene,hgvs_notation,chr,start,end,ref,alt,Ethnic.group,locality,.keep_all = TRUE)

remaining_duplicates<-validated_zlotogora_variants_with_annotation%>%
  ungroup()%>%
  filter(duplicated(validated_zlotogora_variants_with_annotation%>%
                      select(hgvs_notation,population,Ethnic.group,locality)))%>%
  select(hgvs_notation)%>%distinct()

# remove ambiguous variants
validated_zlotogora_variants_with_annotation<-
  validated_zlotogora_variants_with_annotation%>%
  filter(!(hgvs_notation %in% remaining_duplicates$hgvs_notation))

write.table(validated_zlotogora_variants_with_annotation%>%
              select(-c(consequence_terms)),
            file=glue('./output/founder_lists_with_genomic_coordinates/zlotogora_founder_variants_vep.{Sys.Date()}.csv'),
            sep='\t',
            row.names=F,
            quote=F)

```

# generate final table
```{r}
# MOH
moh_raw<-readr::read_delim('/media/SSD/Bioinformatics/Projects/founder_variants_assessment_202204/output/founder_lists_with_genomic_coordinates/preconception_screening_moh_vep.2024-04-15.csv')

moh_proc<-moh_raw%>%
  mutate(source='moh',
         chr=as.character(chr),
         ethnicity=glue('{ethnicity}:{locality}'),
         info=glue('{omim}:{disease}'))%>%
  select(source,original_hgvsc=hgvs_notation,
         hgvsc,cds_ref=ref,cds_alt=alt,strand,chr,start,end,ethnicity,gene,hgvsp,assembly_name,info)
moh_proc

# myscreen
myscreen_raw<-readr::read_delim('/media/SSD/Bioinformatics/Projects/founder_variants_assessment_202204/output/founder_lists_with_genomic_coordinates/preconception_screening_myscreen_vep.2024-04-15.csv')

myscreen_proc<-myscreen_raw%>%
  mutate(source='myscreen',
         chr=as.character(chr),
         ethnicity='.',
         info='.')%>%
  select(source,original_hgvsc=hgvs_notation,
         hgvs_notation,hgvsc,cds_ref=ref,cds_alt=alt,strand,chr,start,end,ethnicity,gene,hgvsp,assembly_name,info)
myscreen_proc

# onco51
onco51_raw<-readr::read_delim('/media/SSD/Bioinformatics/Projects/founder_variants_assessment_202204/output/founder_lists_with_genomic_coordinates/onco51_founder_variants_vep.2024-04-15.csv')

onco51_proc<-onco51_raw%>%
  mutate(source='onco51',
         chr=as.character(chr),
         ethnicity=glue('{sector}:{ethnicity}'),
         info='.')%>%
  select(source,original_hgvsc=hgvs_notation,
         hgvsc,cds_ref=ref,cds_alt=alt,strand,chr,start,end,ethnicity,gene,hgvsp,assembly_name,info)
onco51_proc

# zlotogora
zlotogora_raw<-readr::read_delim('/media/SSD/Bioinformatics/Projects/founder_variants_assessment_202204/output/founder_lists_with_genomic_coordinates/zlotogora_founder_variants_vep.2024-04-15.csv')

# select whether to use only founders
zlotogora_founders_raw<-zlotogora_raw%>%filter(grepl('found|fouder',X.No.of.chromosomes.families,ignore.case=T))

zlotogora_proc<-zlotogora_founders_raw%>%
  mutate(chr=as.character(chr),
         source='imgd',
         ethnicity=glue('{population}:{Ethnic.group}:{locality}'),
         info=glue('{disease.MIM}:{disease};frequency={Carrier.frequency....}'))%>%
  select(source,original_hgvsc=hgvs_notation,
         hgvsc,cds_ref=ref,cds_alt=alt,strand,chr,start,end,ethnicity,gene,hgvsp,assembly_name,info)
zlotogora_proc

final_founder_table<-
  bind_rows(moh_proc,onco51_proc,myscreen_proc,zlotogora_proc)%>%
  filter(!is.na(chr))%>%
  mutate(ref=ifelse(strand==1,cds_ref,chartr('ACGT','TGCA',cds_ref)),
         alt=ifelse(strand==1,cds_alt,chartr('ACGT','TGCA',cds_alt)))%>%
  relocate(ref,alt,.after='end')
write.table(final_founder_table,file=glue('./output/combined_founder_variants_table.{Sys.Date()}.csv'),row.names = F,quote = F,sep='\t')

# fix prob vars
fixed_vars<-final_founder_table%>%slice(c(973,1256, 1257, 1258, 2013, 2073, 2141, 2325, 2326)-1)
write.table(fixed_vars,file=glue('./output/combined_founder_variants_table.missing_vars.{Sys.Date()}.csv'),row.names = F,quote = F,sep='\t')
```

# adding franklin links to zlotogora table

```{r franklin_link}
zlotogora_raw<-readr::read_delim('/media/SSD/Bioinformatics/Projects/founder_variants_assessment_202204/output/founder_lists_with_genomic_coordinates/zlotogora_founder_variants_vep.2024-04-15.csv')

zlotogora_for_franklin<-zlotogora_raw%>%
  rename(cds_ref=ref,cds_alt=alt)%>%
  mutate(ref=ifelse(strand==1,cds_ref,chartr('ACGT','TGCA',cds_ref)),
         alt=ifelse(strand==1,cds_alt,chartr('ACGT','TGCA',cds_alt)))%>%
  relocate(ref,alt,.after='end')
# prep zlotogora table for franklin (to add links)
write.table(zlotogora_for_franklin,            file=glue('./output/zlotogora_founder_variants_vep_for_franklin.{Sys.Date()}.csv'),
            sep='\t',
            row.names=F,
            quote=F)

original_zlotogora<-readr::read_delim('./data/founder_variants_lists/zlotogora_ignd_20240329.csv',delim='\t',comment='')
original_colnames<-colnames(original_zlotogora)
colnames(original_zlotogora)<-make.names(colnames(original_zlotogora))
original_zlotogora<-original_zlotogora%>%
  mutate(Mutation.c=stringr::str_replace_all(Mutation.c,'−','-'),
         Mutation.c=stringr::str_replace_all(Mutation.c,'→','>'),
         Mutation.c=trimws(Mutation.c),
         Mutation.c=gsub("[^\x01-\x7F]", "", Mutation.c))
# this table was generated by franklin based on the zlotogora_for_franklin table
zlotogora_for_franklin_with_franklin_links<-xlsx::read.xlsx2('/media/SSD/Bioinformatics/Projects/founder_variants_assessment_202204/output/franklin/zlotogora_founders_for_franklin.updated_file.2024-05-10.xlsx',sheetIndex = 1)

zlotogora_for_franklin_with_franklin_links<-readxl::read_xlsx('/media/SSD/Bioinformatics/Projects/founder_variants_assessment_202204/output/franklin/zlotogora_founders_for_franklin.updated_file.2024-05-10.xlsx')
zlotogora_for_franklin_with_franklin_links<-zlotogora_for_franklin_with_franklin_links%>%mutate(
  Mutation.c=gsub("[^\x01-\x7F]", "", Mutation.c)
)
# find mismatches with the file we sent franklin
head(zlotogora_for_franklin)
head(zlotogora_for_franklin_with_franklin_links)

comp_data<-zlotogora_for_franklin%>%mutate(start=as.character(start))%>%
  select(Gene,Mutation.c,chr,start,gref=ref,galt=alt,strand)%>%distinct()%>%
  left_join(zlotogora_for_franklin_with_franklin_links%>%select(Gene,Mutation.c,chr,start,ref,alt)%>%distinct())

# variants in franklin file not in original file
in_franklin_not_in_zlotogora<-zlotogora_for_franklin_with_franklin_links%>%filter(!(Mutation.c %in% original_zlotogora$Mutation.c))
in_zlotogora_not_in_franklin<-original_zlotogora%>%filter(!(Mutation.c %in% zlotogora_for_franklin_with_franklin_links$Mutation.c))


comp_data%>%filter(strand==-1,(ref!=gref|alt!=galt))

final_zlotogora_with_links<-original_zlotogora%>%
  left_join(zlotogora_for_franklin_with_franklin_links%>%
              select(Gene,Mutation.c,Mutation..p.,new_link)%>%distinct())

colnames(final_zlotogora_with_links)<-c(original_colnames,'franklin_link')
xlsx::write.xlsx2(data.frame(final_zlotogora_with_links),col.names = T,row.names = F,file='/media/SSD/Bioinformatics/Projects/founder_variants_assessment_202204/output/franklin/table_israel_20240329_from_joel.with_franklin_links.20240513.xlsx')

final_zlotogora_with_links%>%count(!is.na(franklin_link))%>%mutate(rate=n/nrow(final_zlotogora_with_links))
```


# clinvar
```{r oncopanel}
library(vcfR)
library(dplyr)
clinvar_file_vcf<-vcfR::read.vcfR('/media/SSD/Bioinformatics/Databases/clinvar/clinvar_hg38_20231028.plp.vcf')
clinvar_file_vcf_tidy<-vcfR2tidy(clinvar_file_vcf,info_only = T)
clinvar_fixed<-clinvar_file_vcf_tidy$fix%>%separate_rows(GENEINFO,sep = '\\|')%>%
  separate(GENEINFO,into=c('gene_symbol','gene_id'),sep = ':',remove = F)

#onco51_for_vep<-onco51_for_vep%>%filter(gene%in%(bad_vars%>%filter(!is.na(transcript_name))%>%pull(gene)))
query_size<-125

var_chunks<-split(clinvar_fixed$CLNHGVS, ceiling(seq_along(clinvar_fixed$CLNHGVS)/query_size))
chunk_num<-1
clinvar_all_annotated_vars<-NULL

for (hgvs_notations in var_chunks[chunk_num:length(var_chunks)]){
  message(glue('RUNNING CHUNK:{chunk_num}'))
  annotated_vars<-vep_basic_annotation(hgvs_notations)
  #annotated_vars<-validate_vep_variants(annotated_vars,keep=T)
  clinvar_all_annotated_vars<-clinvar_all_annotated_vars%>%bind_rows(annotated_vars)
  chunk_num<-chunk_num+1
}

save(clinvar_all_annotated_vars,file = './output/clinvar_annotated.RData')

all_var_freqs<-NULL
for (var_colocated in clinvar_all_annotated_vars$colocated_variants){
  #print(var_colocated)
  if ('frequencies' %in% colnames(var_colocated)){
    var_freqs<-var_colocated%>%pull(frequencies)
    var_freqs<-var_freqs[[1]]%>%na.omit()
    #print(var_freqs)
    all_var_freqs<-all_var_freqs%>%bind_rows(var_freqs)
  }else{
    all_var_freqs<-all_var_freqs%>%bind_rows(data.frame(af=NA))
  }
}
clinvar_with_freq<-clinvar_fixed%>%left_join(clinvar_all_annotated_vars%>%select(CLNHGVS=input)%>%bind_cols(all_var_freqs))
write.table(clinvar_with_freq,file='/media/SSD/Bioinformatics/Projects/alpha_missense_app_202309/data/clinvar.hg38.plp.with_freq.csv',sep='\t',row.names = F)
```

# Compare variant files
```{r compare_files}
input_files<-c('./output/founder_lists_with_genomic_coordinates/preconception_screening_moh_vep.2023-10-04.withGRCh37.csv',
               './output/founder_lists_with_genomic_coordinates/preconception_screening_myscreen_vep.2023-10-04.csv',
               './output/founder_lists_with_genomic_coordinates/onco51_panel_vep.2023-10-05.csv',
               './output/founder_lists_with_genomic_coordinates/zlotogora_founder_variants_vep.2024-02-29.csv')


for (ifile in input_files){
  fv_data<-readr::read_delim(ifile,delim='\t')
  print(colnames(fv_data))
}

```
