---
title: "vep variant parser"
author: "Ofer Isakov"
date: '2023-03-07'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
setwd('/media/SSD/Bioinformatics/Projects/founder_variants_assessment_202204/')
library(ProjectTemplate)
load.project()
```

```{r aa_conv}
amino_acids <- list(
  A = "Ala",
  R = "Arg",
  N = "Asn",
  D = "Asp",
  C = "Cys",
  E = "Glu",
  Q = "Gln",
  G = "Gly",
  H = "His",
  I = "Ile",
  L = "Leu",
  K = "Lys",
  M = "Met",
  F = "Phe",
  P = "Pro",
  S = "Ser",
  T = "Thr",
  W = "Trp",
  Y = "Tyr",
  V = "Val"
)

# Function to translate shorthand to long format
translate_amino_acids <- function(aa_change,protein_start) {
  # Split the input string by "/" to handle cases like "G/D"
  print(aa_change)
  if (is.na(aa_change)){
    return(data.frame(aa1=NA,aa2=NA,aa_change_long=NA))
  }
  split_sequence <- strsplit(aa_change, split = "/")[[1]]
  
  # Translate each amino acid in the sequence
  translated_sequence <- sapply(split_sequence, function(aa) {
    if (aa %in% names(amino_acids)) {
      return(amino_acids[[aa]])
    } else {
      return(aa) # Return the original character if no match is found
    }
  })
  aa_change_df<-data.frame(aa1=translated_sequence[1],aa2=translated_sequence[2],aa_change_long=glue('p.{translated_sequence[1]}{protein_start}{translated_sequence[2]}'))
  # Combine the translated sequence back into a string with "/"
  return(aa_change_df)
}

```

# MyScreen variants

```{r definitions}

myscreen_variants<-readr::read_delim('./data/founder_variants_lists/davidov_et_al_clinical_genetics_s8.csv',delim='\t')
myscreen_for_vep<-myscreen_variants%>%
  mutate(variant_id=row_number(),gene=trimws(gene_name),variant=trimws(cds_seq),gene='')%>%
  select(-c(gene_name,cds_seq))%>%
  rename(transcript=transcript_name)%>%
  mutate(original_hgvs_notation=glue('{transcript}:{variant}'))

myscreen_for_vep_fixed<-
  myscreen_for_vep%>%
  mutate(hgvs_notation=original_hgvs_notation)%>%
  mutate(transcript=case_when(
    transcript=='NM_001031734.4'~'NM_001397406.1',
    transcript=='NM_000941.3'~'NM_001395413.1',
    transcript=='NM_000060.4'~'NM_001370658.1',
    .default=transcript
  ))
# fix bad transcripts

myscreen_for_vep_annotated<-NULL
for (i in 1:nrow(myscreen_for_vep_fixed)){
#for (i in 1:20){
  message(glue('{i}/{nrow(myscreen_for_vep_fixed)}'))
  variant<-myscreen_for_vep_fixed%>%slice(i)
  variant_position<-vep_cds_to_genomic_coordinates(variant,build='grch37',distinct_genomic_effect = T)
  annotated_variant<-variant%>%left_join(variant_position,by=c('hgvs_notation'))
  myscreen_for_vep_annotated<-myscreen_for_vep_annotated%>%bind_rows(annotated_variant)
}

# failed to parse variants
myscreen_failed_variants<-myscreen_for_vep_annotated%>%filter(!is.na(error))%>%select(variant_id,transcript,gene,hgvs_notation,protein_seq,error)

# find ambiguous variants
myscreen_ambiguous_variant_ids<-myscreen_for_vep_annotated%>%filter(duplicated(variant_id))%>%pull(variant_id)%>%unique()
myscreen_for_vep_annotated$is_ambiguos<-ifelse(myscreen_for_vep_annotated$variant_id%in%myscreen_ambiguous_variant_ids,1,0)

write.table(myscreen_for_vep_annotated%>%select(-consequence_terms),file=glue('./output/founder_lists_with_genomic_coordinates/preconception_screening_myscreen_vep.{Sys.Date()}.csv'),sep='\t',row.names=F,quote=F)

```

# Onco panel 
```{r onco_new}
onco51_variants<-readr::read_delim('./data/founder_variants_lists/onco_panel_51.csv',delim='\t')
onco51_variants<-onco51_variants%>%
  mutate(variant_id=row_number(),
         gene=trimws(gene),variant=trimws(variant),transcript='',
         original_hgvs_notation=glue('{gene}:{variant}'))
onco51_for_vep<-onco51_variants%>%
  mutate(hgvs_notation=original_hgvs_notation)%>%
  filter(hgvs_notation!='BMPR1A:c.-153+13146_*357832del')
  
onco51_for_vep_annotated<-NULL
for (i in 1:nrow(onco51_for_vep)){
  message(glue('{i}/{nrow(onco51_for_vep)}'))
  variant<-onco51_for_vep%>%slice(i)
  variant_position<-vep_cds_to_genomic_coordinates(variant,build='grch37',distinct_genomic_effect = T)
  annotated_variant<-variant%>%left_join(variant_position)
  onco51_for_vep_annotated<-onco51_for_vep_annotated%>%bind_rows(annotated_variant)
}

# validate onco51
# compare start protein
validated_onco51_for_vep_annotated<-onco51_for_vep_annotated%>%
  mutate(original_protein=stringr::str_extract(protein,'p\\..+')%>%stringr::str_replace('\\d+.+',''),
         annotation_protein=stringr::str_extract(hgvsp,'p\\..+')%>%stringr::str_replace('\\d+.+',''))%>%
  mutate(protein_match=ifelse(original_protein==annotation_protein,1,0),
         protein_match=ifelse(is.na(protein_match),0,protein_match))%>%
  group_by(gene,variant,ethnicity,sector)%>%
  slice_max(protein_match)%>%ungroup()
# compare protein notation
validated_onco51_for_vep_annotated<-validated_onco51_for_vep_annotated%>%
  mutate(annotation_protein=stringr::str_extract(hgvsp,'p\\..+'))%>%
  mutate(protein_match=ifelse(protein==annotation_protein,1,0),
         protein_match=ifelse(is.na(protein_match),0,protein_match))%>%
  group_by(gene,variant,ethnicity,sector)%>%
  slice_max(protein_match)%>%ungroup()
# fix deletions
validated_onco51_for_vep_annotated<-validated_onco51_for_vep_annotated%>%
  mutate(hgvs_del_ref=stringr::str_extract(hgvs_notation,'del.+'),
         annotation_del_ref=glue('del{ref}'))%>%
  mutate(del_match=ifelse(hgvs_del_ref==annotation_del_ref,1,0),
         del_match=ifelse(is.na(del_match),0,del_match))%>%
  group_by(gene,variant,ethnicity,sector)%>%
  slice_max(del_match)%>%ungroup()

# grab remaining duplicates
onco51_ambiguous_variant_ids<-validated_onco51_for_vep_annotated%>%filter(duplicated(variant_id))%>%pull(variant_id)%>%unique()
onco51_ambiguous_variants<-validated_onco51_for_vep_annotated%>%filter(variant_id%in%onco51_ambiguous_variant_ids)

# manually select the correct variants
validated_onco51_for_vep_annotated<-validated_onco51_for_vep_annotated%>%
  filter(!(hgvs_notation=='BRCA1:c.5434C>G' & start!=41199693))
  
write.table(validated_onco51_for_vep_annotated,
            file=glue('./output/founder_lists_with_genomic_coordinates/onco51_founder_variants_vep.{Sys.Date()}.csv'),
            sep='\t',
            row.names=F,
            quote=F)

```

# MOH variants 

```{r definitions}

moh_variants<-readr::read_delim('./data/founder_variants_lists/moh_founder_unified.csv',delim='\t')
moh_variants<-moh_variants%>%
  mutate(gene=trimws(gene),variant=trimws(variant),transcript='',
         variant_id=row_number(),
         original_hgvs_notation=glue('{gene}:{variant}'))
# convert problematic gene names
moh_for_vep<-moh_variants%>%
  mutate(
    hgvs_notation=original_hgvs_notation,
    gene = case_when(gene == 'G6PC1' ~ 'G6PC',
                     gene == 'ELP1' ~ 'IKBKAP',
                     gene == 'G6PC1' ~ 'G6PC',
                     gene == 'CRPPA' ~ 'ISPD',
                     gene == 'PJVK' ~ 'DFNB59',
                     gene == 'CCN6' ~ 'WISP3',
                     gene == 'MICOS13' ~ 'C19orf70',
                     !is.na(gene) ~ gene),
    # variant = case_when(
    #   hgvs_notation == 'TK2:c.542T>A' ~ 'TK2:c.635T>A',
    #   hgvs_notation == 'POR:c.1615G>A' ~ 'POR:c.1606G>A',
    #   hgvs_notation == 'CPS1:c.4101+2T>C' ~'CPS1:4134+2T>C',
    #   TRUE ~ hgvs_notation
    hgvs_notation = glue('{gene}:{variant}')
    )
  
moh_for_vep_annotated<-NULL
for (i in 1:nrow(moh_for_vep)){
  message(glue('{i}/{nrow(moh_for_vep)}'))
  variant<-moh_for_vep%>%slice(i)
  variant_position<-vep_cds_to_genomic_coordinates(variant,build='grch37',distinct_genomic_effect = T)
  annotated_variant<-variant%>%left_join(variant_position)
  moh_for_vep_annotated<-moh_for_vep_annotated%>%bind_rows(annotated_variant)
}

# failed variants
moh_failed_variants<-moh_for_vep_annotated%>%filter(!is.na(error))%>%select(variant_id,all_of(colnames(moh_for_vep)),error)
# ambiguous variants
moh_ambiguous_variant_ids<-moh_for_vep_annotated%>%filter(duplicated(variant_id))%>%pull(variant_id)%>%unique()
moh_ambiguous_variants<-moh_for_vep_annotated%>%filter(variant_id%in%moh_ambiguous_variant_ids)

moh_for_vep_annotated$is_ambiguous<-ifelse(moh_for_vep_annotated$variant_id%in%moh_ambiguous_variant_ids,1,0)

write.table(moh_for_vep_annotated,
            file=glue('./output/founder_lists_with_genomic_coordinates/preconception_screening_moh_vep.{Sys.Date()}.csv'),
            sep='\t',
            row.names=F,
            quote=F)

```


# Zlotogora

```{r zlotogora}

#zlotogora_variants<-xlsx::read.xlsx('./data/founder_variants_lists/zlotogora_ingd_202203.xls',sheetIndex = 1)
zlotogora_variants<-readr::read_delim('./data/founder_variants_lists/zlotogora_ignd_20240329.csv',delim='\t',comment='')
colnames(zlotogora_variants)<-make.names(colnames(zlotogora_variants))
# filter only founders
zlotogora_founder_variants<-zlotogora_variants%>%
  filter(grepl('founde',X.No.of.chromosomes.families,ignore.case=T)) # add this if you want to take only the variants that are marked as founder
zlotogora_for_vep<-zlotogora_founder_variants%>%
  mutate(variant_id=row_number(),
         Mutation.c=stringr::str_replace_all(Mutation.c,'−','-'),
         Mutation.c=stringr::str_replace_all(Mutation.c,'→','>'))%>%
  mutate(variant=Mutation.c)%>%
  mutate(transcript='')%>%
  mutate(gene=trimws(Gene),variant=trimws(variant))%>%
  mutate(gene=stringr::str_replace_all(gene,'\\s+',''))%>%
  mutate(variant=stringr::str_replace_all(variant,'‐','-'))%>%
  filter(!is.na(variant))%>%filter(grepl('^c.',variant))%>%
  mutate(variant=stringr::str_replace(variant,'\\;.+',''))%>%
  mutate(variant=stringr::str_replace_all(variant,'\\s+',''),
         variant=stringr::str_replace(variant,'\\(.+',''))%>%
  mutate(gene=case_when(
    gene=='MTP'~'MTTP',
    .default=gene
  ))%>%
  mutate(original_hgvs_notation=glue('{gene}:{variant}'),
         hgvs_notation=original_hgvs_notation)%>%distinct()

zlotogora_for_vep_annotated<-NULL
for (i in 1:nrow(zlotogora_for_vep)){
#for (i in 3171:nrow(zlotogora_for_vep)){
  message(glue('{i}/{nrow(zlotogora_for_vep)}'))
  variant<-zlotogora_for_vep%>%slice(i)
  variant_position<-vep_cds_to_genomic_coordinates(variant,build='grch37',distinct_genomic_effect = T)
  annotated_variant<-variant%>%left_join(variant_position)
  zlotogora_for_vep_annotated<-zlotogora_for_vep_annotated%>%bind_rows(annotated_variant)
}

# grab failed variants
zlotogora_failed<-zlotogora_for_vep_annotated%>%filter(!is.na(error))%>%select(colnames(zlotogora_founder_variants),hgvs_notation,error)

# validate variants
validated_zlotogora_variants_with_annotation<-zlotogora_for_vep_annotated
# if there is a location, validate it
validated_zlotogora_variants_with_annotation<-validated_zlotogora_variants_with_annotation%>%
  mutate(annotation_location=glue('chr{chr}:{start}'))%>%
  mutate(location_match=ifelse(location==annotation_location,1,0),
         location_match=ifelse(is.na(location_match),0,location_match))%>%
  group_by(gene,variant,Ethnic.group,locality)%>%
  slice_max(location_match)%>%ungroup()

# compare annotation protein with variant annotation
validated_zlotogora_variants_with_annotation<-validated_zlotogora_variants_with_annotation%>%
  mutate(annotation_protein=stringr::str_extract(hgvsp,'p\\..+'))%>%
  mutate(protein_match=ifelse(Mutation..p.==annotation_protein,1,0),
         protein_match=ifelse(is.na(protein_match),0,protein_match))%>%
  group_by(gene,variant,Ethnic.group,locality)%>%
  slice_max(protein_match)%>%ungroup()

# ambiguous variants
zlotogora_ambiguous_variant_ids<-validated_zlotogora_variants_with_annotation%>%filter(duplicated(variant_id))%>%pull(variant_id)%>%unique()
zlotogora_ambiguous_variants<-validated_zlotogora_variants_with_annotation%>%filter(variant_id%in%zlotogora_ambiguous_variant_ids)

validated_zlotogora_variants_with_annotation$is_ambiguous<-ifelse((validated_zlotogora_variants_with_annotation$variant_id %in% zlotogora_ambiguous_variant_ids),1,0)
  
write.table(validated_zlotogora_variants_with_annotation,
            file=glue('./output/founder_lists_with_genomic_coordinates/zlotogora_founder_variants_vep.{Sys.Date()}.csv'),
            sep='\t',
            row.names=F,
            quote=F)

```

# Wide screen MOH

```{r moh_wide_screen}
moh_ws<-xlsx::read.xlsx2('./data/founder_variants_lists/moh/moh_07_07_24.xlsx',
                         sheetIndex = 1)
# moh_ws$X.<-NULL
# colnames(moh_ws)<-c('disease_id','disease_name','omim','gene','omim_id','transcript','variant','ethnicity')
moh_ws<-
  moh_ws%>%mutate(across(everything(), ~ stringr::str_trim(.)))%>%
  #separate_rows(variant,sep = '/')%>%
  mutate(variant_id=row_number(),
         variant = c_mutation,
         original_hgvs_notation=ifelse(transcript!='' & !is.na(transcript),
                                       glue('{transcript}:{variant}'),
                                       glue('{gene}:{variant}')))

# Fix the input variant list (if variants are written incorrectly)
moh_ws_fixed<-
  moh_ws%>%
  mutate(variant=stringr::str_replace(variant,'\\,',''))%>%
  # mutate(variant=ifelse(variant=='c.415(-1)G>A','c.415-1G>A',variant))%>%
  # mutate(variant=stringr::str_replace(variant,'(\\s|\\.)*(\\.|\\;|\\,)$',''))%>%
  # mutate(variant=stringr::str_replace(variant,'\\<','\\>'))%>%
  mutate(variant=stringr::str_replace(variant,'c\\.\\.','c\\.'))%>%
  # mutate(variant=stringr::str_replace(variant,'c\\.\\s+','c\\.'))%>%
  # mutate(variant=stringr::str_replace(variant,'\\s+\\>','\\>'))%>%
  # mutate(variant=stringr::str_replace(variant,'\\s*\\(.+',''))%>%
  # mutate(variant=stringr::str_replace(variant,'^\\-','c\\.\\-'))%>%
  # mutate(gene=case_when(
  #   gene=='C8ORF37'~'C8orf37',
  #   gene=='DHDPSL (HOGA1)'~'HOGA1',
  #   gene=='GNPTAG'~'GNPTG',
  #   gene=='IL7RA'~'IL7R',
  #   gene=='CRPPA (ISPD)'~'ISPD',
  #   gene=='TBCB (CKAPI)'~'TBCB',
  #   transcript=='NM_005055.5'~'RAPSN',
  #   .default=gene
  # ))%>%
  # mutate(variant=case_when(
  #   variant=='c.788InsC'~'c.788dup',
  #   variant=='c.1034insT'~'c.1034dup',
  #   variant=='g.22736_29335delinsGGCACCTG'~'c.169-4949_352+1467delinsGGCACCTG',
  #   .default=variant
  # ))%>%
  # mutate(transcript=case_when(
  #   transcript=='NM_000098'~'NM_000098.3',
  #   transcript=='NM_013335'~'NM_013335.4',
  #   transcript=='NM_000275'~'NM_000275.3',
  #   gene=='MICOS13'~'NM_205767.3',
  #   gene=='MMUT'~'NM_000255.4',
  #   gene=='PJVK'~'NM_001042702.5',
  #   transcript=='NM_005055.5'~'',
  #   .default=transcript
  # ))%>%
  mutate(hgvs_notation=ifelse(transcript=='',
                              glue('{gene}:{variant}'),
                              glue('{transcript}:{variant}')))

moh_ws_vep_annotated<-NULL
for (i in 1:nrow(moh_ws_fixed)){
#for (i in 1:20){
  message(glue('{i}/{nrow(moh_ws_fixed)}'))
  variant<-moh_ws_fixed%>%slice(i)
  variant_position<-vep_cds_to_genomic_coordinates(variant,build='grch37',distinct_genomic_effect=TRUE)
  annotated_variant<-variant%>%left_join(variant_position,by=c('hgvs_notation'='hgvs_notation'))
  moh_ws_vep_annotated<-moh_ws_vep_annotated%>%bind_rows(annotated_variant)
}

# grab all ambiguous variants
ambiguous_variant_ids<-moh_ws_vep_annotated%>%filter(duplicated(variant_id))%>%pull(variant_id)%>%unique()
moh_ws_vep_annotated$is_ambiguous<-ifelse(moh_ws_vep_annotated$variant_id%in%ambiguous_variant_ids,1,0)

write.table(moh_ws_vep_annotated,
            file=glue('./output/founder_lists_with_genomic_coordinates/moh_ws_vep.{Sys.Date()}.csv'),
            sep='\t',
            row.names=F,
            quote=F)
```


# Analyze MOH ws

```{r analyze_moh_ws}
moh_ws_ambiguous_table<-
  moh_ws_vep_annotated%>%
  filter(variant_id%in%ambiguous_variant_ids)

# compare with full founders list
cols_for_comp<-c('original_hgvs_notation','chr','start','end','genomic_ref','genomic_alt','gene_hgvs','transcript_hgvs','is_ambiguous','source')
moh_ws_for_merge<-moh_ws_vep_annotated%>%
  mutate(cds_change=stringr::str_extract(hgvsc,'c\\..+'),
         transcript=stringr::str_replace(hgvsc,':.+',''),
         gene_hgvs=glue('{gene}:{cds_change}'),
         transcript_hgvs=glue('{transcript}:{cds_change}'),
         source='moh_ws')%>%
  select(all_of(cols_for_comp))
founders_for_merge<-readr::read_delim('./output/combined_founder_variants_table.2024-07-12.csv')%>%
  mutate(original_hgvs_notation=original_hgvsc,
         cds_change=stringr::str_extract(hgvsc,'c\\..+'),
         transcript=stringr::str_replace(hgvsc,':.+',''),
         gene_hgvs=glue('{gene_symbol}:{cds_change}'),
         transcript_hgvs=glue('{transcript}:{cds_change}'))%>%
  select(all_of(cols_for_comp))

# comparison by location
moh_ws_for_merge<-moh_ws_for_merge%>%
  left_join(founders_for_merge%>%
              mutate(location_found=1)%>%
              distinct(chr,start,end,genomic_ref,genomic_alt,location_found,is_ambiguous))%>%
  mutate(location_found=ifelse(is.na(location_found),0,1))

founders_for_merge<-founders_for_merge%>%
  left_join(moh_ws_for_merge%>%
              mutate(location_found=1)%>%
              distinct(chr,start,end,genomic_ref,genomic_alt,location_found,is_ambiguous))%>%
  mutate(location_found=ifelse(is.na(location_found),0,1))

# comparison by transcript
moh_ws_for_merge<-moh_ws_for_merge%>%
  mutate(transcript_found=ifelse((transcript_hgvs%in%founders_for_merge$transcript_hgvs)|
                                       (gene_hgvs%in%founders_for_merge$gene_hgvs),1,0))%>%
  mutate(location_or_transcript_found=location_found+transcript_found)

founders_for_merge<-founders_for_merge%>%
  mutate(transcript_found=ifelse((transcript_hgvs%in%moh_ws_for_merge$transcript_hgvs)|
                                       (gene_hgvs%in%moh_ws_for_merge$gene_hgvs),1,0))%>%
  mutate(location_or_transcript_found=location_found+transcript_found)

merged_list<-moh_ws_for_merge%>%bind_rows(founders_for_merge)
merged_list<-merged_list%>%
  mutate(comparison=case_when(
    location_or_transcript_found>0 & source=='moh_ws'~'moh_ws_found_in_list',
    location_or_transcript_found==0 & source=='moh_ws'~'moh_ws_not_found_in_list',
    location_or_transcript_found>0 & source!='moh_ws'~'founder_in_moh_ws',
    location_or_transcript_found==0 & source!='moh_ws'~'founder_not_in_moh_ws',
    ))

xlsx::write.xlsx2(moh_ws_ambiguous_table%>%
                    select(original_hgvs_notation,
                           original_transcript=transcript,
                           hgvs_notation,
                           transcript_id,
                           chr,start,end,ref,alt,
                           hgvsp)%>%
                    as.data.frame(),
                  file=glue('./output/moh_ws_comparison_table.{Sys.Date()}.xlsx'),
                  sheetName = 'ambiguous_variants',
                  row.names = F)

xlsx::write.xlsx2(merged_list%>%
                    as.data.frame(),
                  file=glue('./output/moh_ws_comparison_table.{Sys.Date()}.xlsx'),
                  sheetName = 'moh_ws_vs_full_list',
                  row.names = F,
                  append = T)

  
```

# generate final table
```{r}
# Raw tables
## MOH (old screen)
moh_raw<-readr::read_delim('/media/SSD/Bioinformatics/Projects/founder_variants_assessment_202204/output/founder_lists_with_genomic_coordinates/preconception_screening_moh_vep.2024-07-10.csv')
## myscreen
myscreen_raw<-readr::read_delim('/media/SSD/Bioinformatics/Projects/founder_variants_assessment_202204/output/founder_lists_with_genomic_coordinates/preconception_screening_myscreen_vep.2024-07-10.csv')
## onco51
onco51_raw<-readr::read_delim('/media/SSD/Bioinformatics/Projects/founder_variants_assessment_202204/output/founder_lists_with_genomic_coordinates/onco51_founder_variants_vep.2024-07-10.csv')
## zlotogora
zlotogora_raw<-readr::read_delim('/media/SSD/Bioinformatics/Projects/founder_variants_assessment_202204/output/founder_lists_with_genomic_coordinates/zlotogora_founder_variants_vep.2024-07-11.csv')

moh_proc<-moh_raw%>%
  mutate(source='moh',
         chr=as.character(chr),
         ethnicity=glue('{ethnicity}:{locality}'),
         info=glue('{omim}:{disease}'))%>%
  select(source,original_hgvsc=hgvs_notation,
         hgvsc,cds_ref=ref,cds_alt=alt,strand,chr,start,end,genomic_ref,genomic_alt,ethnicity,gene_symbol,transcript,hgvsp,assembly_name,info,is_ambiguous,error)
moh_proc

myscreen_proc<-myscreen_raw%>%
  mutate(source='myscreen',
         chr=as.character(chr),
         ethnicity='.',
         info='.')%>%
  select(source,original_hgvsc=hgvs_notation,
         hgvs_notation,hgvsc,cds_ref=ref,cds_alt=alt,strand,chr,start,end,genomic_ref,genomic_alt,ethnicity,gene_symbol,transcript,hgvsp,assembly_name,info,error)
myscreen_proc

onco51_proc<-onco51_raw%>%
  mutate(source='onco51',
         chr=as.character(chr),
         ethnicity=glue('{sector}:{ethnicity}'),
         info='.')%>%
  select(source,original_hgvsc=hgvs_notation,
         hgvsc,cds_ref=ref,cds_alt=alt,strand,chr,start,end,genomic_ref,genomic_alt,ethnicity,gene_symbol,transcript,hgvsp,assembly_name,info,error)
onco51_proc


# select whether to use only founders
zlotogora_founders_raw<-zlotogora_raw%>%filter(grepl('found|fouder',X.No.of.chromosomes.families,ignore.case=T))

zlotogora_proc<-zlotogora_founders_raw%>%
  mutate(chr=as.character(chr),
         source='imgd',
         ethnicity=glue('{population}:{Ethnic.group}:{locality}'),
         info=glue('{disease.MIM}:{disease};frequency={Carrier.frequency....}'))%>%
  select(source,original_hgvsc=hgvs_notation,
         hgvsc,cds_ref=ref,cds_alt=alt,strand,chr,start,end,genomic_ref,genomic_alt,ethnicity,gene_symbol,transcript,hgvsp,assembly_name,info,is_ambiguous,error)
zlotogora_proc

final_founder_table<-
  bind_rows(moh_proc,onco51_proc,myscreen_proc,zlotogora_proc)%>%
  filter(!is.na(chr))%>%
  filter(is.na(error))%>%
  filter(is.na(is_ambiguous)|is_ambiguous!=1)
write.table(final_founder_table,file=glue('./output/combined_founder_variants_table.{Sys.Date()}.csv'),row.names = F,quote = F,sep='\t')


```

# adding franklin links to zlotogora table

```{r franklin_link}
zlotogora_raw<-readr::read_delim('/media/SSD/Bioinformatics/Projects/founder_variants_assessment_202204/output/founder_lists_with_genomic_coordinates/zlotogora_founder_variants_vep.2024-04-15.csv')

zlotogora_for_franklin<-zlotogora_raw%>%
  rename(cds_ref=ref,cds_alt=alt)%>%
  mutate(ref=ifelse(strand==1,cds_ref,chartr('ACGT','TGCA',cds_ref)),
         alt=ifelse(strand==1,cds_alt,chartr('ACGT','TGCA',cds_alt)))%>%
  relocate(ref,alt,.after='end')
# prep zlotogora table for franklin (to add links)
write.table(zlotogora_for_franklin,            file=glue('./output/zlotogora_founder_variants_vep_for_franklin.{Sys.Date()}.csv'),
            sep='\t',
            row.names=F,
            quote=F)

original_zlotogora<-readr::read_delim('./data/founder_variants_lists/zlotogora_ignd_20240329.csv',delim='\t',comment='')
original_colnames<-colnames(original_zlotogora)
colnames(original_zlotogora)<-make.names(colnames(original_zlotogora))
original_zlotogora<-original_zlotogora%>%
  mutate(Mutation.c=stringr::str_replace_all(Mutation.c,'−','-'),
         Mutation.c=stringr::str_replace_all(Mutation.c,'→','>'),
         Mutation.c=trimws(Mutation.c),
         Mutation.c=gsub("[^\x01-\x7F]", "", Mutation.c))
# this table was generated by franklin based on the zlotogora_for_franklin table
zlotogora_for_franklin_with_franklin_links<-xlsx::read.xlsx2('/media/SSD/Bioinformatics/Projects/founder_variants_assessment_202204/output/franklin/zlotogora_founders_for_franklin.updated_file.2024-05-10.xlsx',sheetIndex = 1)

zlotogora_for_franklin_with_franklin_links<-readxl::read_xlsx('/media/SSD/Bioinformatics/Projects/founder_variants_assessment_202204/output/franklin/zlotogora_founders_for_franklin.updated_file.2024-05-10.xlsx')
zlotogora_for_franklin_with_franklin_links<-zlotogora_for_franklin_with_franklin_links%>%mutate(
  Mutation.c=gsub("[^\x01-\x7F]", "", Mutation.c)
)
# find mismatches with the file we sent franklin
head(zlotogora_for_franklin)
head(zlotogora_for_franklin_with_franklin_links)

comp_data<-zlotogora_for_franklin%>%mutate(start=as.character(start))%>%
  select(Gene,Mutation.c,chr,start,gref=ref,galt=alt,strand)%>%distinct()%>%
  left_join(zlotogora_for_franklin_with_franklin_links%>%select(Gene,Mutation.c,chr,start,ref,alt)%>%distinct())

# variants in franklin file not in original file
in_franklin_not_in_zlotogora<-zlotogora_for_franklin_with_franklin_links%>%filter(!(Mutation.c %in% original_zlotogora$Mutation.c))
in_zlotogora_not_in_franklin<-original_zlotogora%>%filter(!(Mutation.c %in% zlotogora_for_franklin_with_franklin_links$Mutation.c))


comp_data%>%filter(strand==-1,(ref!=gref|alt!=galt))

final_zlotogora_with_links<-original_zlotogora%>%
  left_join(zlotogora_for_franklin_with_franklin_links%>%
              select(Gene,Mutation.c,Mutation..p.,new_link)%>%distinct())

colnames(final_zlotogora_with_links)<-c(original_colnames,'franklin_link')
xlsx::write.xlsx2(data.frame(final_zlotogora_with_links),col.names = T,row.names = F,file='/media/SSD/Bioinformatics/Projects/founder_variants_assessment_202204/output/franklin/table_israel_20240329_from_joel.with_franklin_links.20240513.xlsx')

final_zlotogora_with_links%>%count(!is.na(franklin_link))%>%mutate(rate=n/nrow(final_zlotogora_with_links))
```


# clinvar
```{r oncopanel}
library(vcfR)
library(dplyr)
clinvar_file_vcf<-vcfR::read.vcfR('/media/SSD/Bioinformatics/Databases/clinvar/clinvar_hg38_20231028.plp.vcf')
clinvar_file_vcf_tidy<-vcfR2tidy(clinvar_file_vcf,info_only = T)
clinvar_fixed<-clinvar_file_vcf_tidy$fix%>%separate_rows(GENEINFO,sep = '\\|')%>%
  separate(GENEINFO,into=c('gene_symbol','gene_id'),sep = ':',remove = F)

#onco51_for_vep<-onco51_for_vep%>%filter(gene%in%(bad_vars%>%filter(!is.na(transcript_name))%>%pull(gene)))
query_size<-125

var_chunks<-split(clinvar_fixed$CLNHGVS, ceiling(seq_along(clinvar_fixed$CLNHGVS)/query_size))
chunk_num<-1
clinvar_all_annotated_vars<-NULL

for (hgvs_notations in var_chunks[chunk_num:length(var_chunks)]){
  message(glue('RUNNING CHUNK:{chunk_num}'))
  annotated_vars<-vep_basic_annotation(hgvs_notations)
  #annotated_vars<-validate_vep_variants(annotated_vars,keep=T)
  clinvar_all_annotated_vars<-clinvar_all_annotated_vars%>%bind_rows(annotated_vars)
  chunk_num<-chunk_num+1
}

save(clinvar_all_annotated_vars,file = './output/clinvar_annotated.RData')

all_var_freqs<-NULL
for (var_colocated in clinvar_all_annotated_vars$colocated_variants){
  #print(var_colocated)
  if ('frequencies' %in% colnames(var_colocated)){
    var_freqs<-var_colocated%>%pull(frequencies)
    var_freqs<-var_freqs[[1]]%>%na.omit()
    #print(var_freqs)
    all_var_freqs<-all_var_freqs%>%bind_rows(var_freqs)
  }else{
    all_var_freqs<-all_var_freqs%>%bind_rows(data.frame(af=NA))
  }
}
clinvar_with_freq<-clinvar_fixed%>%left_join(clinvar_all_annotated_vars%>%select(CLNHGVS=input)%>%bind_cols(all_var_freqs))
write.table(clinvar_with_freq,file='/media/SSD/Bioinformatics/Projects/alpha_missense_app_202309/data/clinvar.hg38.plp.with_freq.csv',sep='\t',row.names = F)
```

# Compare variant files
```{r compare_files}
input_files<-c('./output/founder_lists_with_genomic_coordinates/preconception_screening_moh_vep.2023-10-04.withGRCh37.csv',
               './output/founder_lists_with_genomic_coordinates/preconception_screening_myscreen_vep.2023-10-04.csv',
               './output/founder_lists_with_genomic_coordinates/onco51_panel_vep.2023-10-05.csv',
               './output/founder_lists_with_genomic_coordinates/zlotogora_founder_variants_vep.2024-02-29.csv')


for (ifile in input_files){
  fv_data<-readr::read_delim(ifile,delim='\t')
  print(colnames(fv_data))
}

```
